File: Awesome.AI.Source/Awesome.AI/Common/GPTVector6D.cs
namespace Awesome.AI.Common
{
    public struct GPTVector6D
    {
        public double xx { get; set; }
        public double yy { get; set; }
        public double zz { get; set; }
        public double ww { get; set; }
        public double vv { get; set; }
        public double uu { get; set; }

        public double magnitude
        {
            get
            {
                return Math.Sqrt(
                    xx * xx +
                    yy * yy +
                    zz * zz +
                    ww * ww +
                    vv * vv +
                    uu * uu);
            }
        }

        public GPTVector6D(
            double x,
            double y,
            double z,
            double w,
            double v,
            double u)
        {
            xx = x;
            yy = y;
            zz = z;
            ww = w;
            vv = v;
            uu = u;
        }

        public GPTVector6D Add(GPTVector6D v1, GPTVector6D v2)
        {
            return new GPTVector6D(
                v1.xx + v2.xx,
                v1.yy + v2.yy,
                v1.zz + v2.zz,
                v1.ww + v2.ww,
                v1.vv + v2.vv,
                v1.uu + v2.uu);
        }

        public GPTVector6D Sub(GPTVector6D v1, GPTVector6D v2)
        {
            return new GPTVector6D(
                v1.xx - v2.xx,
                v1.yy - v2.yy,
                v1.zz - v2.zz,
                v1.ww - v2.ww,
                v1.vv - v2.vv,
                v1.uu - v2.uu);
        }

        public GPTVector6D Mul(GPTVector6D vector, double scalar)
        {
            return new GPTVector6D(
                vector.xx * scalar,
                vector.yy * scalar,
                vector.zz * scalar,
                vector.ww * scalar,
                vector.vv * scalar,
                vector.uu * scalar);
        }

        public GPTVector6D Div(GPTVector6D vector, double scalar)
        {
            if (scalar == 0)
                throw new DivideByZeroException("Cannot divide by zero.");

            return new GPTVector6D(
                vector.xx / scalar,
                vector.yy / scalar,
                vector.zz / scalar,
                vector.ww / scalar,
                vector.vv / scalar,
                vector.uu / scalar);
        }

        public double Dot(GPTVector6D v1, GPTVector6D v2)
        {
            return
                v1.xx * v2.xx +
                v1.yy * v2.yy +
                v1.zz * v2.zz +
                v1.ww * v2.ww +
                v1.vv * v2.vv +
                v1.uu * v2.uu;
        }

        public GPTVector6D Unit()
        {
            double len = magnitude;

            if (len == 0)
                throw new DivideByZeroException(
                    "Cannot normalize a zero vector.");

            return new GPTVector6D(
                xx / len,
                yy / len,
                zz / len,
                ww / len,
                vv / len,
                uu / len);
        }

        public GPTVector6D Reverse()
        {
            return new GPTVector6D(
                100.0d - xx,
                100.0d - yy,
                100.0d - zz,
                100.0d - ww,
                100.0d - vv,
                100.0d - uu);
        }

        public GPTVector6D ReverseUnit()
        {
            return new GPTVector6D(
                -xx,
                -yy,
                -zz,
                -ww,
                -vv,
                -uu);
        }

        public double DistanceTo(GPTVector6D other)
        {
            double dx = xx - other.xx;
            double dy = yy - other.yy;
            double dz = zz - other.zz;
            double dw = ww - other.ww;
            double dv = vv - other.vv;
            double du = uu - other.uu;

            return Math.Sqrt(
                dx * dx +
                dy * dy +
                dz * dz +
                dw * dw +
                dv * dv +
                du * du);
        }
    }
}

File: Awesome.AI.Python/Common/GPTVector6D.py
from __future__ import annotations
from dataclasses import dataclass, fields
import math


@dataclass
class GPTVector6D:
    xx: float = 0.0
    yy: float = 0.0
    zz: float = 0.0
    ww: float = 0.0
    vv: float = 0.0
    uu: float = 0.0

    def _values(self): return tuple(getattr(self, f.name) for f in fields(self))
    @classmethod
    def _from(cls, values): return cls(*values)
    def Add(self, v1, v2): return self._from(a + b for a, b in zip(v1._values(), v2._values()))
    def Sub(self, v1, v2): return self._from(a - b for a, b in zip(v1._values(), v2._values()))
    def Mul(self, vector, scalar): return self._from(a * scalar for a in vector._values())
    def Div(self, vector, scalar):
        if scalar == 0: raise ZeroDivisionError("scalar")
        return self._from(a / scalar for a in vector._values())
    def Dot(self, v1, v2): return sum(a * b for a, b in zip(v1._values(), v2._values()))
    @property
    def magnitude(self): return math.sqrt(self.Dot(self, self))
    def Unit(self):
        length = self.magnitude
        if length == 0: raise ZeroDivisionError("Cannot normalize a zero vector.")
        return self.Div(self, length)
    def Reverse(self): return self._from(100.0 - value for value in self._values())
    def ReverseUnit(self): return self._from(-value for value in self._values())
    def DistanceTo(self, other): return self.Sub(self, other).magnitude
